### ISSUE TITLE
[Asset|Document|Object] retry transaction if delete() throws RetryableException

### ISSUE THREAD

**Issue by brusch** [pimcore/pimcore#8089]
Similar to `save()`: 

https://github.com/pimcore/pimcore/blob/master/models/DataObject/AbstractObject.php#L680

**Comment by weisswurstkanone** [comment]
I think that is not that easy as it sounds (compared to an update). 

"Delete" also kills the children and grandchildren etc, snowball effect literally. 
So there is no way to do this in an atomic way and restore them if some piece in the process fails. 

Nevertheless, I would proceed as suggested as it would be a good improvement.

@brusch Wdyt ?

**Comment by brusch** [comment]
@weisswurstkanone agree, please proceed

**Comment by brusch** [comment]
#9758 needs to be solved first. The thing is that children are getting deleted as well during the transaction, which can't be atomic in the current implementation: 
https://github.com/pimcore/pimcore/blob/5f9c86682639403f22475bf5dc11758d962ba9fd/models/Asset.php#L1005
https://github.com/pimcore/pimcore/blob/c6263b315cb06e75037e9484aeed9635cf8feb6d/models/Document.php#L766
https://github.com/pimcore/pimcore/blob/0f680f7c71f858936e016ba552faf8de79339cc9/models/DataObject/AbstractObject.php#L581

Marking the children just as deleted in the database and delete them later on in the message handler would get us around the problem.

**Comment by kingjia90** [comment]
Had a quick look while working on the other mentioned issue, at least for assets, shouldn't it be already working with the [transaction nesting](https://www.doctrine-project.org/projects/doctrine-dbal/en/4.1/reference/transactions.html#transaction-nesting)? 

[File deletion](https://github.com/pimcore/pimcore/blob/ba7de58efb4f9a3062e343d6bd2474d7a8f7ce5d/models/Asset.php#L1012-L1019) happens only after the main transaction `commit()` is successful (which implies that all the nested save points never got rollback)


My concern is maybe that it would be retrying 5 times for each element, maybe could be too much, maybe could be reduce by adding something like
```
            $maxRetries = $isNested ? 3 : 5;
            for ($retries = 0; $retries < $maxRetries; $retries++) {
```

Update: after some more digging, i think it might be tricky to skip or rollback the thumbnails, so we have likely to wait for the other issue to be fully resolved
https://github.com/pimcore/pimcore/blob/dce29a1e855f443733465c3b3b38b1a5bac78407/models/Asset.php#L1041-L1044

**Comment by mattamon** [comment]
@kingjia90 this issue was closed https://github.com/pimcore/pimcore/issues/9758 so actually this ticket cannot be solved anymore right or can we moved forward here?

**Comment by kingjia90** [comment]
@mattamon thank you for checking it, indeed, this ticket would be technically blocked indefinitely due to the mentioned issue not being possibly solved first or any time soon. 
It is now to be tackled at an architectural level and need some more discussion and brainstorming

**Comment by brusch** [comment]
@kingjia90 Just had a look at documents, assets and objects and I agree, that it should work already as you have mentioned [here](https://github.com/pimcore/pimcore/issues/8089#issuecomment-2298187989). 

Maybe we should then also consider the `bool $isNested = false` parameter used for assets for documents and objects. Might be necessary to know if a delete is nested or not.

**Comment by kingjia90** [comment]
Thank you @brusch for having a look

I have opened a draft PR  https://github.com/pimcore/pimcore/pull/17853 , for docs and objects seems fine without `$isNested`.
If i got it correctly, `isNested` was kept for a the Assets as a way to unify and "postpone" the mass physical file deletion on success of every child deletion.

I have the following "blockers":

- The same problem on rolling back files seen on https://github.com/pimcore/pimcore/issues/17534, more specifically about the thumbnail https://github.com/pimcore/pimcore/blob/11.4/models/Asset.php#L1015 getting cleared even on failed deletion which cannot be restored by the rollback. Tried in the PR to first collect all the paths and ids and perform it by `clearThumbnailsByPaths()` after the db commit are successful but only works for direct children at the moment and trying to find a way to recursively collect every element. Is it on the right path? 
- Should it be tied to RetryableException, DeadlockException or be more generic?

Thank you in advance

**Comment by brusch** [comment]
@kingjia90 I think for thumbnails it's fine if we loose them, they are temp files and can be treated as such. 
It's anyway just in case of an error, but we need to ensure the consistency of the original source files. 

I'd stay on `RetryableException` as it is now already for data objects, it includes `DeadlockException` and having something more generic is imho currently not necessary. If we see the need anytime, we can of course extend it to consider more types of exceptions.

**Comment by cancan101** [comment]
I do think this line in `save`:
https://github.com/pimcore/pimcore/blob/769575dc84e888c095b608faa4c1b81ba0bf32d3/models/Asset.php#L552
should be handling `RetryableException` based on: https://www.doctrine-project.org/projects/doctrine-dbal/en/4.2/reference/transactions.html#error-handling.
